libdnf/utils/string.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright (C) 2020-2021 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#include "string.hpp"
+
+#include <iterator>
+#include <sstream>
+
+
+namespace libdnf::utils::string {
+
+
+std::string join(const std::vector<std::string> & vec, const std::string & delimiter) {
```

Comment:
Good idea. However, this only works with `std::vector<std::string>`. I prefer a general tool. Which works e.g. with `const char * []`, or `std::forward_list<char *>`, ... So something that works with "iterable" containers that contain elements implicitly convertible to `std::string`.

Idea:
```
/// Join all items in a container into a string using a separator
template<typename ContainerT>
inline std::string join(const ContainerT & input, const std::string & separator) {
    auto it = std::begin(input);
    auto it_end = std::end(input);

    std::string result;

    // Append first element
    if (it != it_end) {
        result.append(*it);
        ++it;
    }

    for (; it != it_end; ++it) {
        result.append(separator);
        result.append(*it);
    }

    return result;
}
```

And tests e.g.:
```
CPPUNIT_ASSERT_EQUAL(std::string("aa, bb"), join(std::vector<std::string>{"aa", "bb"}, ", "));
CPPUNIT_ASSERT_EQUAL(std::string("aa, bb"), join(std::forward_list<std::string>{"aa", "bb"}, ", "));
CPPUNIT_ASSERT_EQUAL(std::string("aa, bb"), join(std::forward_list<const char *>{"aa", "bb"}, ", "));
```